analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs

```diff
@@ -26,4 +28,167 @@ public abstract class RestrictDeserializedTypesBase : SymbolicRuleCheck
+                ? SerializationConstraint.Safe
+                : SerializationConstraint.Unsafe;
+            if (constraint == SerializationConstraint.Unsafe)
+            {
+                additionalLocationsForOperations[instance] = bindToTypeDeclaration;
+            }
+            state = state.SetOperationConstraint(instance, constraint);
+
+            if (instance.TrackedSymbol() is { } symbol)
+            {
+                if (constraint == SerializationConstraint.Unsafe)
+                {
+                    additionalLocationsForSymbols[symbol] = bindToTypeDeclaration;
+                }
+                state = state.SetSymbolConstraint(symbol, constraint);
+            }
+            return state;
+        }
+        return null;
+    }
+
+    private static IOperation BinderAssignmentInstance(ProgramState state, IAssignmentOperationWrapper assignment) =>
+        state.ResolveCaptureAndUnwrapConversion(assignment.Target).AsPropertyReference() is { Property.Name: "Binder", Instance: { } propertyInstance }
+        && propertyInstance.Type.IsAny(FormattersWithBinder)
+            ? state.ResolveCaptureAndUnwrapConversion(propertyInstance)
+            : null;
+
+    private bool BinderIsSafe(ProgramState state, IAssignmentOperationWrapper assignment, out SyntaxNode bindToTypeDeclaration)
+    {
+        if (state[assignment.Value]?.HasConstraint(ObjectConstraint.Null) is true)
+        {
+            bindToTypeDeclaration = null;
+            return false;
+        }
+        else
+        {
+            bindToTypeDeclaration = DeclarationCandidates(assignment.Value).FirstOrDefault(IsBindToType);
+            return bindToTypeDeclaration is null || ThrowsOrReturnsNull(bindToTypeDeclaration);
+        }
+    }
+
+    private IEnumerable<SyntaxNode> DeclarationCandidates(IOperation operation) =>
+        SemanticModel.GetTypeInfo(operation.Syntax).Type?.DeclaringSyntaxReferences.SelectMany(x => x.GetSyntax().DescendantNodes());
+
+    private SyntaxNode AdditionalLocation(ProgramState state, IOperation operation)
+    {
+        operation = state.ResolveCaptureAndUnwrapConversion(operation);
+        return additionalLocationsForOperations.TryGetValue(operation, out var methodDeclaration)
+            || (operation.TrackedSymbol() is { } symbol && additionalLocationsForSymbols.TryGetValue(symbol, out methodDeclaration))
+            ? methodDeclaration
+            : null;
+    }
+
+    private static IInvocationOperationWrapper? UnsafeDeserialization(ProgramState state, IOperation operation) =>
+        operation.AsInvocation() is { } invocation
+        && invocation.TargetMethod.Name == "Deserialize"
+        && invocation.TargetMethod.ContainingType.IsAny(TypesWithDeserializeMethod)
+        && state[invocation.Instance]?.HasConstraint(SerializationConstraint.Unsafe) is true
+            ? invocation : null;
+
+    protected abstract bool IsBindToType(SyntaxNode methodDeclaration);
```

Comment:
Move the abstract methods [after the constant and field declarations](https://github.com/SonarSource/sonar-dotnet/blob/master/docs/coding-style.md#class-members).